charms/reactive/patterns/request_response.py

```diff
@@ -128,6 +128,20 @@ def _load(cls, request_sources):
                 request.response = cls.RESPONSE_CLASS._load(request)
                 cls._cache[request.request_id] = request
 
+    @classmethod
+    def _load_app(cls, relations):
+        for source in relations:
+            if source.application_name is None:
+                continue
+            app_requests = source.to_publish_app
+            for key, app_request_data in app_requests.items():
+                if not key.startswith('request_'):
+                    continue
+                source.app = True
+                request = cls(source, source['request_id'])
+                request.response = cls.RESPONSE_CLASS._load(request)
```

Comment:
The response class is also going to need a `_load_app` method which will need to read the `received_app` collection suggested above.

Reply:
> The response class is also going to need a `_load_app` method which will need to read the `received_app` collection suggested above.

BaseResponse's _load just returns response do i need to make _load_app inside BaseReponse class? and how can i call received_app in this _load_app if I make it?

`   
    @classmethod
    def _load(cls, request):
        response = cls(request)
        if response._key not in response._source_data:
            return None  # no response found
        else:
            return response
`

Reply:
The `_load` method on both `BaseRequest` and `BaseResponse` return instances of the class from the data source.  The question is where that data comes from, and there are four possibilities each for both `RequestBase` and `ResponseBase`:

  * `unit.received` for data coming from a single unit on the remote side of the relation
  * `relation.received_app` for data coming from the leader of the remote side of the unit to represent the entire app (new in this PR)
  * `unit.to_publish` for data that this unit published for itself only
  * `relation.to_publish_app` for data that the leader of this app published for this entire app (new in this PR, and can only be read or written by the leader)

The original set of classes was written assuming that the data would only ever come in over unit-level data (since app-level data wasn't available then). It also was (poorly, by me) designed to try to track whether the data was being read or published (hence the `is_received` and `_source_data` attributes), so needs to handle both cases.  (Doing it again, the classes would just be de/serializers for a given data bag, and the relation class would be responsible for keeping track of which data bag a given request or response needs to be associated with, since it's much easier to do cleanly at that level.)

All that said, the issue is that you've added a way to read request data from the app-level relation data, but you haven't added a way to write request data to the app-level relation data nor have you added a way to either read or write response data from/to app-level relation data.